Automattic/woocommerce-payments
# Fix/handle async failed refunds

Fixes #2285 

#### Changes proposed in this Pull Request

At the moment when a WCPay refund fails and the webhook request is received, we update the order meta `_wcpay_refund_status` to `failed`. However, the refund is not deleted. This PR enables fetching the last WC Refund and saving the WCPay refund id as meta, which is then used to delete WC Refunds when the failed webhook request is received.

This PR also updates the refund order notes to include the WCPay refund id and changing the order status to `failed` if the order is fully refunded. For partially refunded orders, the status is not changed.

#### Testing instructions

* Follow the steps mentioned in #2285 to replicate the problem.
* Switch to the branch `fix/handle-async-failed-refunds`.
* The changes can be tested locally, but the Jurassic Tube tunnel needs to be enabled to receive webhook requests.
* Create a new order from front-end with the card `4000 0000 0000 5126`
* Fully refund the order from WP-ADMIN and confirm that the order note added includes the WCPay refund id - `re_xxxxxxxxxxxxxxxxxxxxxxxx`
* Refresh the page to see the new order note indicating refund failure.
* Confirm that the refund create is no longer present on the order and the order status is changed to `Failed`
* Create another order using the same card details above and do a partial refund.
* Confirm that the refund create is no longer present on the order and the order status is the same as before.

-------------------


**Post merge**
